atef/widgets/config/utils.py

```diff
@@ -900,3 +903,193 @@ def get_info(self) -> Dict[str, Any]:
+
+
+def combine_results(results: List[Result]) -> Result:
+    """
+    Combines results into a single result.
+
+    Takes the highest severity, and currently all the reasons
+    """
+    severity = util.get_maximum_severity([r.severity for r in results])
+    reason = str([r.reason for r in results]) or ''
+
+    return Result(severity=severity, reason=reason)
+
+
+class ConfigTreeModel(QtCore.QAbstractItemModel):
+    """
+    Item model for tree data.  Goes through all this effort due to the need for
+    tooltips, icons, etc.  This model is READ-ONLY, and does not implement
+    the ``setData`` method.
+
+    Expects the item to specifically TreeItem, which each holds a
+    Configuration or Comparison
+    """
+    def __init__(self, *args, data: TreeItem, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.tree_data = data or TreeItem()
+        self.root_item = self.tree_data
+        self.headers = ['Name', 'Status', 'Type']
+
+    def headerData(
+        self,
+        section: int,
+        orientation: Qt.Orientation,
+        role: int
+    ) -> Any:
+        """
+        Returns the header data for the model.
+        Currently only displays horizontal header data
+        """
+        if role != Qt.DisplayRole:
+            return
+
+        if orientation == Qt.Horizontal:
+            return self.headers[section]
+
+    def index(self, row: int, column: int, parent: QtCore.QModelIndex = None):
+        """ Returns the index of the item in the model """
+        if not self.hasIndex(row, column, parent):
+            return QtCore.QModelIndex()
+
+        parent_item = None
+        if not parent or not parent.isValid():
+            parent_item = self.root_item
+        else:
+            parent_item = parent.internalPointer()
+
+        child_item = parent_item.child(row)
+        if child_item:
+            return self.createIndex(row, column, child_item)
+
+        # all else
+        return QtCore.QModelIndex()
+
+    def parent(self, index: QtCore.QModelIndex):
+        """ Returns the parent of the given model item. """
+        if not index.isValid():
+            return QtCore.QModelIndex()
+        child = index.internalPointer()
+        parent = child.parent()
+        if parent == self.root_item:
+            return QtCore.QModelIndex()
+
+        return self.createIndex(parent.row(), 0, parent)
+
+    def rowCount(self, parent: QtCore.QModelIndex):
+        """ Called by tree view to determine number of children an item has. """
+        if not parent.isValid():
+            parent_item = self.root_item
+        else:
+            parent_item = parent.internalPointer()
+        return parent_item.childCount()
+
+    def columnCount(self, parent: QtCore.QModelIndex):
+        if not parent.isValid():
+            parent_item = self.root_item
+        else:
+            parent_item = parent.internalPointer()
+        return parent_item.columnCount()
+
+    def data(self, index, role):
```

Comment:
Annotations?

Reply:
Apparently I was all over the place... This method's missing annotations and the one above it is missing a docstring... 🤦